### PR TITLE
feat: set `Http::application_id` in ready event handler

### PR DESCRIPTION
**Pull request by vaporoxx**
This PR modifies `Shard` so that it overwrites the HTTP application id when it receives a ready payload. This is useful so that you do not have to save the application id in the environment too instead of just the token. I'm not sure if it is the best implementation but I have tested it and it seems to work.

### CODE REVIEW

File: src/http/client.rs
Comment (kangalio) [inline src/http/client.rs:269]: All 25 callers of this function have `.ok_or(HttpError::ApplicationIdMissing)`. Maybe this function should return `Result<u64, HttpError>` directly?

**Comment by kangalio** [comment]
Nice PR, this is pretty useful.

1. Could you fix the compilation errors in the doc tests?
2. This PR makes an Http instance mandatory for using the gateway, not sure if that's ok
3. Serenity has an ApplicationId struct, might be nice to use this here instead of raw u64

File: src/http/client.rs
Comment (vaporoxx) [reply src/http/client.rs:269]: I'm not sure if that makes sense from a user perspective since the method doesn't return any other error variants, maybe a new private method would fit better?

Comment (kangalio) [reply src/http/client.rs:269]: Good idea. Name suggestions: `try_application_id`, `application_id_or_err`

**Comment by vaporoxx** [comment]
> This PR makes an Http instance mandatory for using the gateway, not sure if that's ok

That's a reasonable concern (although the docs strongly advise against creating a new shard manually), I'll wait for maintainer feedback on this before updating all the doc tests

> Serenity has an ApplicationId struct, might be nice to use this here instead of raw u64

The `Http` client uses raw numbers everywhere, I think changing it to `ApplicationId` would be too inconsistent

**Comment by nickelc** [comment]
@vaporox I rebased your PR onto the current `next` branch.

**Review by kangalio — APPROVED** [review]

**Review by arqunis — APPROVED** [review]

**Pull request merged**
